src/main/java/bio/terra/stairway/FlightMap.java

```diff
@@ -77,33 +102,78 @@ public void makeImmutable() {
    * @throws ClassCastException if found, not castable to the requested type
    */
   public <T> T get(String key, Class<T> type) {
-    Object value = map.get(key);
+    String value = map.get(key);
     if (value == null) {
       return null;
     }
 
-    if (type.isInstance(value)) {
-      return type.cast(value);
+    try {
+      return getObjectMapper().readValue(value, type);
+    } catch (final JsonProcessingException ex) {
+      throw new ClassCastException(
+          "Found value '" + value + "' is not an instance of type " + type.getName());
     }
-    throw new ClassCastException(
-        "Found value '" + value.toString() + "' is not an instance of type " + type.getName());
   }
 
+  /**
+   * Gets the raw (serialized) data stored for an Object in the hash map.
+   *
+   * @param key Key to lookup the data for in the map.
+   * @return String mapped to key, or null if key does not exist.
+   */
+  public String getRaw(String key) {
+    return map.get(key);
+  }
+
+  /**
+   * Serialize an Object and place it into the map at a given key.
+   *
+   * @param key Key to store the object under.
+   * @param value Object to serialize and store.
+   * @throws JsonConversionException if object could not be serialized.
+   */
   public void put(String key, Object value) {
+    try {
+      map.put(key, getObjectMapper().writeValueAsString(value));
+    } catch (final JsonProcessingException ex) {
+      throw new JsonConversionException("Failed to convert object to json string", ex);
+    }
+  }
+
+  /**
+   * Stores a raw (serialized) data value for a given key.
+   *
+   * @param key Key to store the data under.
+   * @param value Data to store.
+   */
+  public void putRaw(String key, String value) {
     map.put(key, value);
   }
 
   public String toJson() {
     try {
-      return getObjectMapper().writeValueAsString(map);
-    } catch (JsonProcessingException ex) {
+      // Serializing as Map<String, RawJsonValue> will produce a format compatible with older code's
+      // Map<String, Object> representation.
+      Map<String, RawJsonValue> temporaryMap = new HashMap<>();
```

Comment:
Not if the older code doesn't have the RawJsonValue class.If we're trying to prevent rollback/concurrent version problems, this doesn't work. 
Consider what's at head as version N and this PR merged as version N + 1.  version N can't read what version N + 1 wrote, as it won't know what a `RawJsonValue` is.
Am I misunderstanding how the custom serializer will write this out?

I thought the intention was to allow this PR to be rolled back if necessary. We need to write just in the old version format to do that. Unless you want to do a preemptory PR introducing `RawJsonValue`.

Reply:
Or if it does create something that's backwards compatible with version N, why bother with this RawJsonValue and just use the version N serialization?

Reply:
The reason for mapping to `RawJsonValue` class is that we can then inject a custom serializer that gets hooked for type `RawJsonValue`, which injects the raw JSON string (including the object type).  That's the highest layer in Jackson where we can hook raw JSON; upper layers will escape it and treat it as a string, lower layers require us to roll the entire map.

Reply:
Reverted per previous comment.  Next phase will remove toJson().